File: src/main/cc/wfa/virtual_people/core/model/utils/population_node_helper.cc
// Copyright 2026 The Cross-Media Measurement Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "wfa/virtual_people/core/model/utils/population_node_helper.h"

#include <cstdint>
#include <memory>
#include <utility>
#include <vector>

#include "absl/status/status.h"
#include "absl/status/statusor.h"
#include "absl/strings/str_cat.h"
#include "absl/strings/string_view.h"
#include "common_cpp/macros/macros.h"
#include "wfa/virtual_people/common/label.pb.h"
#include "wfa/virtual_people/common/model.pb.h"
#include "wfa/virtual_people/core/model/utils/distributed_consistent_hashing.h"

namespace wfa_virtual_people {

absl::Status CollapseQuantumLabel(const QuantumLabel& quantum_label,
                                  absl::string_view seed_suffix,
                                  PersonLabelAttributes& output_label) {
  if (quantum_label.labels_size() == 0) {
    return absl::InvalidArgumentError("Empty quantum label.");
  }
  if (quantum_label.labels_size() != quantum_label.probabilities_size()) {
    return absl::InvalidArgumentError(absl::StrCat(
        "The sizes of labels and probabilities are different in quantum label ",
        quantum_label.DebugString()));
  }
  std::vector<DistributionChoice> distribution;
  distribution.reserve(quantum_label.probabilities_size());
  for (int i = 0; i < quantum_label.probabilities_size(); ++i) {
    distribution.emplace_back(
        DistributionChoice({i, quantum_label.probabilities(i)}));
  }
  ASSIGN_OR_RETURN(
      std::unique_ptr<DistributedConsistentHashing> hashing,
      DistributedConsistentHashing::Build(std::move(distribution)));
  int32_t index = hashing->Hash(absl::StrCat(
      "quantum-label-collapse-", quantum_label.seed(), seed_suffix));
  output_label.MergeFrom(quantum_label.labels(index));
  return absl::OkStatus();
}

}  // namespace wfa_virtual_people

File: modules/virtual-people-core-serving/src/main/kotlin/org/wfanet/virtualpeople/core/model/utils/PopulationNodeHelper.kt
// Copyright 2026 The Cross-Media Measurement Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package org.wfanet.virtualpeople.core.model.utils

import com.google.common.hash.Hashing
import java.nio.charset.StandardCharsets
import org.wfanet.virtualpeople.common.PersonLabelAttributes
import org.wfanet.virtualpeople.common.QuantumLabel

/**
 * Shared utilities for PopulationNodeImpl and RankedPopulationNodeImpl.
 *
 * Keeps seed computation and quantum label collapsing in sync across both node types.
 */
object PopulationNodeHelper {

  /** Compute the VID selection seed from a random seed and acting fingerprint. */
  fun computeVidSeed(randomSeed: String, actingFingerprint: Long): ULong {
    return Hashing.farmHashFingerprint64()
      .hashString("$randomSeed${actingFingerprint.toULong()}", StandardCharsets.UTF_8)
      .asLong()
      .toULong()
  }

  /**
   * Collapse the [quantumLabel] to a single label based on the probabilities, and merge to
   * [outputLabel].
   */
  fun collapseQuantumLabel(
    quantumLabel: QuantumLabel,
    seedSuffix: String,
    outputLabel: PersonLabelAttributes.Builder
  ) {
    if (quantumLabel.labelsCount == 0) {
      error("Empty quantum label.")
    }
    if (quantumLabel.labelsCount != quantumLabel.probabilitiesCount) {
      error("The sizes of labels and probabilities are different in quantum label. $quantumLabel")
    }
    val distribution =
      quantumLabel.probabilitiesList.mapIndexed { index, it -> DistributionChoice(index, it) }
    val hashing = DistributedConsistentHashing(distribution)
    val index = hashing.hash("quantum-label-collapse-${quantumLabel.seed}$seedSuffix")
    outputLabel.mergeFrom(quantumLabel.getLabels(index))
  }
}
